osu.Game/Screens/Select/FilterQueryParser.cs

```diff
@@ -19,8 +19,31 @@ public static class FilterQueryParser
             @"\b(?<key>\w+)(?<op>(:|=|(>|<)(:|=)?))(?<value>("".*""[!]?)|(\S*))",
             RegexOptions.Compiled | RegexOptions.IgnoreCase);
 
+        private static readonly Regex difficulty_query_syntax_regex = new Regex(
+            @"(\s|^)((\[(?>\[(?<level>)|[^[\]]+|\](?<-level>))*(?(level)(?!))\](\s|$))|(\[.*))",
+            RegexOptions.Compiled);
+
         internal static void ApplyQueries(FilterCriteria criteria, string query)
         {
+            while (true)
```

Comment:
This has to be rewritten to avoid using `while(true)`. It's just too dangerous.

I also wonder if regex should even be used for this as it obfuscates things and is likely not adding much function.

Reply:
Sorry for the delay. I did it without the Regex now, let me know if you think that this is a better solution.

Reply:
I'm not sure. Neither solution is really appealing. Will wait for another opinion.